cache/redis_cache.go

```diff
@@ -94,88 +86,227 @@ func (r *redisCache) nbOfBytes() uint64 {
+	// since the cached results in redis are too big, we can't fetch all of them because of the memory overhead.
+	// We will create an io.reader that will fetch redis bulk by bulk to reduce the memory usage.
+	// But before that, since the usage of the reader could take time and the object in redis could disappear btw 2 fetchs
+	// we need to make sure the TTL will be long enough to avoid nasty side effects
+	// nb: it would be better to retry the flow if such a failure happened but this require a huge refactoring of proxy.go
+	if ttl <= 5*time.Second {
 		return nil, ErrMissing
 	}
+
+	reader := NewRedisStreamReader(uint64(offset), r.client, stringKey, metadata.Length)
 	value := &CachedData{
-		ContentMetadata: ContentMetadata{
-			Length:   payload.Length,
-			Type:     payload.Type,
-			Encoding: payload.Encoding,
-		},
-		Data: bytes.NewReader(decoded),
-		Ttl:  ttl,
+		ContentMetadata: *metadata,
+		Data:            &ioReaderDecorator{Reader: reader},
+		Ttl:             ttl,
 	}
 
 	return value, nil
 }
 
-func (r *redisCache) Put(reader io.Reader, contentMetadata ContentMetadata, key *Key) (time.Duration, error) {
-	data, err := toBytes(reader)
-	if err != nil {
-		return 0, err
-	}
+// this struct is here because CachedData requires an io.ReadCloser
+// but logic in the the Get function generates only an io.Reader
+type ioReaderDecorator struct {
+	io.Reader
+}
 
-	encoded := base64.StdEncoding.EncodeToString(data)
-	payload := &redisCachePayload{
-		Length: contentMetadata.Length, Type: contentMetadata.Type, Encoding: contentMetadata.Encoding, Payload: encoded,
-	}
+func (m ioReaderDecorator) Close() error {
+	return nil
+}
+func (r *redisCache) encodeString(s string) []byte {
+	n := uint32(len(s))
 
-	marshalled, err := json.Marshal(payload)
-	if err != nil {
-		return 0, nil
+	b := make([]byte, 0, n+4)
+	b = append(b, byte(n>>24), byte(n>>16), byte(n>>8), byte(n))
+	b = append(b, s...)
+	return b
+}
+
+func (r *redisCache) decodeString(bytes []byte) (string, int) {
+	b := bytes[:4]
+	n := uint32(b[3]) | (uint32(b[2]) << 8) | (uint32(b[1]) << 16) | (uint32(b[0]) << 24)
+	s := bytes[4 : 4+n]
+	return string(s), int(4 + n)
+}
+
+func (r *redisCache) encodeMetadata(contentMetadata *ContentMetadata) []byte {
+	cLength := contentMetadata.Length
+	cType := r.encodeString(contentMetadata.Type)
+	cEncoding := r.encodeString(contentMetadata.Encoding)
+	b := make([]byte, 0, len(cEncoding)+len(cType)+8)
+	b = append(b, byte(cLength>>56), byte(cLength>>48), byte(cLength>>40), byte(cLength>>32), byte(cLength>>24), byte(cLength>>16), byte(cLength>>8), byte(cLength))
+	b = append(b, cType...)
+	b = append(b, cEncoding...)
+	return b
+}
+
+func (r *redisCache) decodeMetadata(b []byte) (*ContentMetadata, int) {
+	cLength := uint64(b[7]) | (uint64(b[6]) << 8) | (uint64(b[5]) << 16) | (uint64(b[4]) << 24) | uint64(b[3])<<32 | (uint64(b[2]) << 40) | (uint64(b[1]) << 48) | (uint64(b[0]) << 56)
```

Comment:
I think we miss error handling. What if malicious actor tampers payloads in redis? It would make chproxy panic. We could handle that more securely by ignoring cached data. WDYT?